File: android/app/src/main/kotlin/com/example/book_reader/plugin/JsEvalUtilsPlugin.java
package com.example.book_reader.plugin;

import android.content.Context;
import com.example.book_reader.common.GlobalVal;
import com.example.book_reader.common.LogUtil;
import java.util.HashMap;
import javax.script.SimpleBindings;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;

public class JsEvalUtilsPlugin implements MethodCallHandler, FlutterPlugin {

    private static final String TAG = "JsEvalUtilsPlugin";
    private Context applicationContext;
    private MethodChannel methodChannel;
    private Result tmpResult;
    private Boolean hasWrite = false;

    /** Plugin registration. */
    public static void registerWith(PluginRegistry.Registrar registrar) {
        final JsEvalUtilsPlugin instance = new JsEvalUtilsPlugin();
        instance.onAttachedToEngine(registrar.context(), registrar.messenger());
    }

    @Override
    public void onAttachedToEngine(FlutterPluginBinding binding) {
        onAttachedToEngine(binding.getApplicationContext(), binding.getBinaryMessenger());
    }

    private void onAttachedToEngine(Context applicationContext, BinaryMessenger messenger) {
        this.applicationContext = applicationContext;
        methodChannel = new MethodChannel(messenger, "js_eval_plugin");
        methodChannel.setMethodCallHandler(this);
    }

    @Override
    public void onDetachedFromEngine(FlutterPluginBinding binding) {
        applicationContext = null;
        methodChannel.setMethodCallHandler(null);
        methodChannel = null;
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        tmpResult = result;
        switch (call.method) {
            case "evalJs":
                String paramsJsCode = (call.argument("jsCode") == null) ? "" : call.argument("jsCode").toString();
                Object paramsResult = call.argument("result");
                String paramsBaseUrl = (call.argument("baseUrl") == null) ? "" : call.argument("baseUrl").toString();
                String paramsSearchPage = (call.argument("searchPage") == null) ? "" : call.argument("searchPage").toString();
                String paramsSearchKey = (call.argument("searchKey") == null) ? "" : call.argument("searchKey").toString();
                String paramsTfAjaxContentKey = (call.argument("tfAjaxContentKey") == null) ? "" : call.argument("tfAjaxContentKey").toString();
                hasWrite = false;

                SimpleBindings bindings = new SimpleBindings();
                bindings.put("java", this);
                bindings.put("result", paramsResult);
                bindings.put("baseUrl", paramsBaseUrl);
                bindings.put("searchPage", paramsSearchPage);
                bindings.put("searchKey", paramsSearchKey);

                String[] tfAjaxContentList = paramsTfAjaxContentKey.split("-");
                for (String str : tfAjaxContentList) {
                    String key = "tfAjaxContent" + str;
                    String value = (call.argument(key) == null) ? "" : call.argument(key).toString();
                    bindings.put(key, value);
                }
                try {
                    Object retObj = GlobalVal.SCRIPT_ENGINE.eval(paramsJsCode, bindings);
                    if (!hasWrite) result.success(retObj);
                } catch (Exception e) {
                    if (!hasWrite) result.error(TAG, e.getMessage(), e);
                }
                break;
            default:
                result.notImplemented();
                break;
        }
    }


    public String ajax(String postUrl) {
        LogUtil.i(TAG, "Ajax =================== " + postUrl);
        HashMap<String, String> map = new HashMap<String, String>();
        map.put("ReEvalKey", "javaAjax");
        map.put("url", postUrl);
        if(!hasWrite) tmpResult.success(map);
        hasWrite = true;
        return "";
    }

    public String put(String key, String value) {
        LogUtil.i(TAG, "Put =================== " + key + " =========== " + value);
        HashMap<String, String> map = new HashMap<String, String>();
        map.put("ReEvalKey", "javaPut");
        map.put("key", key);
        map.put("value", value);
        if (!hasWrite) tmpResult.success(map);
        hasWrite = true;
        return "";
    }

    public String get(String key) {
        LogUtil.i(TAG, "Get =================== " + key);
        HashMap<String, String> map = new HashMap<String, String>();
        map.put("ReEvalKey", "javaGet");
        map.put("key", key);
        if (!hasWrite) tmpResult.success(map);
        hasWrite = true;
        return "";
    }

    public String base64Decoder(String base64) {
        HashMap<String, Object> map = new HashMap<String, Object>();
        map.put("ReEvalKey", "javaBase64Decoder");
        map.put("base64", base64);
        if (!hasWrite) tmpResult.success(map);
        hasWrite = true;
        return "";
    }

    public String setContent(Object html) {
        HashMap<String, Object> map = new HashMap<String, Object>();
        map.put("ReEvalKey", "javaSetContent");
        map.put("html", html);
        if (!hasWrite) tmpResult.success(map);
        hasWrite = true;
        return "";
    }

    public String getString(String rule) {
        HashMap<String, Object> map = new HashMap<String, Object>();
        map.put("ReEvalKey", "javaGetString");
        map.put("rule", rule);
        if (!hasWrite) tmpResult.success(map);
        hasWrite = true;
        return "";
    }

    public String getStringList(String rule) {
        HashMap<String, Object> map = new HashMap<String, Object>();
        map.put("ReEvalKey", "javaGetStringList");
        map.put("rule", rule);
        if (!hasWrite) tmpResult.success(map);
        hasWrite = true;
        return "";
    }

    public String getElements(String rule) {
        HashMap<String, Object> map = new HashMap<String, Object>();
        map.put("ReEvalKey", "javaGetElements");
        map.put("rule", rule);
        if (!hasWrite) tmpResult.success(map);
        hasWrite = true;
        return "";
    }
}

File: ios/Runner/Plugin/JsEvalUtilsPlugin.m
#import "JsEvalUtilsPlugin.h"
#import <JavaScriptCore/JavaScriptCore.h>

@implementation JsEvalUtilsPlugin

static JSContext *jsContext;
static bool isJSContextInit = false;
//static bool isHasWrite = false;

//用于输出完整长度日志
#define NSLog(format, ...) printf("TIME:%s FILE:%s(%d行) FUNCTION:%s \n %s\n\n",__TIME__, [[[NSString stringWithUTF8String:__FILE__] lastPathComponent] UTF8String], __LINE__, __PRETTY_FUNCTION__, [[NSString stringWithFormat:(format), ##__VA_ARGS__] UTF8String])

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel methodChannelWithName:@"js_eval_plugin" binaryMessenger:[registrar messenger]];
    JsEvalUtilsPlugin* instance = [[JsEvalUtilsPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([@"evalJs" isEqualToString:call.method]) {
        @try{
            NSString *paramsJsCode = call.arguments[@"jsCode"];
            NSString *paramsResult = call.arguments[@"result"];
            NSString *paramsBaseUrl = call.arguments[@"baseUrl"];
            NSString *paramsSearchPage = call.arguments[@"searchPage"];
            NSString *paramsSearchKey = call.arguments[@"searchKey"];
            NSString *paramsTfAjaxContentKey = call.arguments[@"tfAjaxContentKey"];
            NSString *paramsTfGetStringListKey = call.arguments[@"tfGetStringListKey"];
            NSString *paramsTfGetElementsKey = call.arguments[@"tfGetElementsKey"];
            //初始化引擎
            if(!isJSContextInit){
                jsContext = [[JSContext alloc] init];
                isJSContextInit = true;
            }
            __block bool isHasWrite = false;
            //变量赋值
            jsContext[@"result"] = paramsResult;
            jsContext[@"baseUrl"] = paramsBaseUrl;
            jsContext[@"searchPage"] = paramsSearchPage;
            jsContext[@"searchKey"] = paramsSearchKey;
            
            if(paramsTfAjaxContentKey.length > 0){
                NSArray *tfAjaxContentList = [paramsTfAjaxContentKey componentsSeparatedByString:@"-"];
                for (id str in tfAjaxContentList) {
                    NSString *key = [@"tfAjaxContent" stringByAppendingString:str];
                    jsContext[key] = call.arguments[key];
                }
            }
            if(paramsTfGetStringListKey.length > 0){
                NSArray *tfGetStringListList = [paramsTfGetStringListKey componentsSeparatedByString:@"-"];
                for (id str in tfGetStringListList) {
                    NSString *key = [@"tfGetStringList" stringByAppendingString:str];
                    jsContext[key] = call.arguments[key];
                }
            }
            if(paramsTfGetElementsKey.length > 0){
                NSArray *tfGetElementsList = [paramsTfGetElementsKey componentsSeparatedByString:@"-"];
                for (id str in tfGetElementsList) {
                    NSString *key = [@"tfGetElements" stringByAppendingString:str];
                    jsContext[key] = call.arguments[key];
                }
            }
            NSLog(@"JS调用规则：%@", paramsJsCode);
            
            //JS 调用OC函数
            jsContext[@"javaAjax"] = ^(JSValue *getUrl) {
                NSLog(@"javaAjax====== %@",getUrl);
                NSDictionary *dict = @{@"ReEvalKey":@"javaAjax", @"url": getUrl.toObject};
                if(!isHasWrite) result(dict);
                isHasWrite = true;
                return getUrl;
            };
            jsContext[@"javaPut"] = ^(JSValue *key, JSValue *value) {
                NSLog(@"Put====== %@ ===== %@", key, value);
                NSDictionary *dict = @{@"ReEvalKey":@"javaPut", @"key": key.toObject, @"value": value.toObject};
                if(!isHasWrite) result(dict);
                isHasWrite = true;
                return key;
            };
            jsContext[@"javaGet"] = ^(JSValue *key) {
                NSLog(@"Get====== %@",key);
                NSDictionary *dict = @{@"ReEvalKey":@"javaGet", @"key": key.toObject};
                if(!isHasWrite) result(dict);
                isHasWrite = true;
                return key;
            };
            jsContext[@"javaBse64Decoder"] = ^(JSValue *base64) {
                NSDictionary *dict = @{@"ReEvalKey":@"javaBse64Decoder", @"base64": base64.toObject};
                if(!isHasWrite) result(dict);
                isHasWrite = true;
                return base64;
            };
            jsContext[@"javaSetContent"] = ^(JSValue *html) {
                NSDictionary *dict = @{@"ReEvalKey":@"javaSetContent", @"html": html.toObject};
                if(!isHasWrite) result(dict);
                isHasWrite = true;
                return html;
            };
            jsContext[@"javaGetString"] = ^(JSValue *rule) {
                NSDictionary *dict = @{@"ReEvalKey":@"javaGetString", @"rule": rule.toObject};
                if(!isHasWrite) result(dict);
                isHasWrite = true;
                return rule;
            };
            jsContext[@"javaGetStringList"] = ^(JSValue *rule) {
                NSDictionary *dict = @{@"ReEvalKey":@"javaGetStringList", @"rule": rule.toObject};
                if(!isHasWrite) result(dict);
                isHasWrite = true;
                return rule;
            };
            jsContext[@"javaGetElements"] = ^(JSValue *rule) {
                NSDictionary *dict = @{@"ReEvalKey":@"javaGetElements", @"rule": rule.toObject};
                if(!isHasWrite) result(dict);
                isHasWrite = true;
                return rule;
            };
            
            JSValue * value = [jsContext evaluateScript:paramsJsCode];
            //如果是undefined，则返回原值
            if(value == nil || [value.toString isEqual: @"undefined"]){
                NSLog(@"JS调用返回结果为空：");
                if(!isHasWrite) result(paramsResult);
            }
            else{
                if(!isHasWrite) result(value.toObject);
            }
        }
        @catch (NSException *exception){
            result(@"参数转换失败");
        }
    } else {
        result(FlutterMethodNotImplemented);
    }
}

@end
